modules/tensor_mechanics/test/src/kernels/HomogenizedTotalLagrangianStressDivergenceS.C

```diff
@@ -0,0 +1,259 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "HomogenizedTotalLagrangianStressDivergenceS.h"
+
+// MOOSE includes
+#include "Function.h"
+#include "MooseVariableScalar.h"
+
+registerMooseObject("TensorMechanicsTestApp", HomogenizedTotalLagrangianStressDivergenceS);
```

Comment:
There are already the standard Tensor Mechanics homogenization classes, and I didn't want to remove those as I can't say yet if this new way is definitively "better" (i.e. runs faster; since the equations being solved are identical). So, I'm including them in the TestApp as an example of how to use the scalar augmentation class but keep all the existing homogenization routines in place.
Namely, any current users of the homogenization classes would need to revise all of their input files to remove the extra objects that aren't used anymore.

Reply:
What do those suffix "A" "S" mean? Do you have any documentation for those classes?

Reply:
I added some more sentences about the suffixes S, A and R. The documentation is in the header files; basically, the Jacobian of these systems is large and coupled (3 spatial variables and 2 scalar variables), so there are variable ways to partition which terms get computed within which instance of the class. Those instances are controlled by parameters "component" and "prime scalar" in the input file, which map to "alpha" and "beta" within the code.

The purpose of these files is to show that while multiple spatial and scalar variables are possible using the base class, the developer and user need to be careful in the implementation. Namely, the outer loops outside of Kernel already loop over all possible coupled variables; it is the job of the developer to ensure they know which current coupled variable they are giving contributions to.

This first question also ties to the second. The point of the "augmentation" system is to provide built-in interfaces (that the developer can ignore the implementation of) to do spatial integration for residual/Jacobian contributions of objects coupled to scalar variables.

I'll encourage you to read (modules/tensor_mechanics/test/include/kernels/HomogenizedTotalLagrangianStressDivergenceR.h), which has some matrices and vectors spelled out. But generically, say I have a Kernel which has a "focus" spatial variable "uy" and "focus" scalar variable "e1" (which has 3 components) and 2 coupled spatial variables "ux uz" and coupled scalar variable "e2" having 6 components. Then the "scalar augmentation system" provides standardized Qp routines to call for computing and assembling all the following components of R and J:
R =
[R_uy
,R_e_1]
J  = 
[J_uy_ux, J_uy_uy, J_uy_uz, J_uy_e1, J_uy_e2
,J_e1_ux, J_e1_uy, J_e1_uz, J_e1_e1, J_e1_e2]
In the current MOOSE Kernel object, you can only do the following if you hack the code:
R =
[R_uy]
J  = 
[J_uy_ux, J_uy_uy, J_uy_uz, J_uy_e1, J_uy_e2
,000000, J_uy_e1, 000000, 000000, 000000]
The other e1 row entries have to get handled by calls to Scalar Kernels and user objects.

Nonetheless, the developer might not want every instance of that Kernel to compute something for all 10 of those entries, depending how they want the object to be referenced in the input file. Since {x,y,z} is different numbering than {1,2}, the user can't really pair up x with1 and y with 2 and then be left with z. So instead they might want 5 Kernels in the input file, 3 doing spatial contributions and 2 doing scalar contributions. Exactly analogous to how a user has 3 Tensor Mechanics stress divergence Kernels to handle the 3 displacement components. And you can automate that with an Action system to hide/automate the creation of all the Kernels for standard cases.

Hope this helps; and if you have an idea for making some of this "pretty", and it is currently not coming through in one of the doc files, then suggest where to put it.

Reply:
@ttruster  this is very helpful information. thank you.